README.md
 A [Sass]/[PostCSS] toolbelt for superspeed front-end development

[PostCSS]: https://github.com/postcss/postcss
[Gulp]: https://github.com/gulpjs/gulp
[Sass]: http://sass-lang.com


## Installation

```js
npm install postcss-styler
```

## Example
Documentation coming soon...

CHANGELOG.md
# 2.0.6 - 2017-10-11
* Fixed issue with values when not found

# 2.0.5 - 2017-10-11
* Fixed issues with latest update

# 2.0.4 - 2017-10-11
* Fixed issue with icons not working together with var

# 2.0.3 - 2017-09-29
* Updated dependecies

# 2.0.2 - 2017-08-23
* Added better detection for variable and color functions
* Updated dependencies

# 2.0.1 = 2017-07-21
* Added color module

# 2.0.0 - 2017-07-19
* Initial release

index.js
/**
 * POSTCSS STYLER
 * @version 2.0.6
 * @author Arpad Hegedus <hegedus.arpad@gmail.com>
 */

let postcss = require('postcss');

module.exports = postcss.plugin('styler', (opt = {}) => {
    let css = postcss();
    opt = Object.assign({
        modules: ['remove-comments', 'media', 'at-style', 'icon', 'color', 'var', 'google-color', 'font', 'font-vsize', 'sides', 'timing-function', 'clear-fix', 'morphicon', 'gridder'],
        sass: {
            indentType: 'tab',
            indentWidth: '1',
            outputStyle: 'expanded'
        },
        grid: {
            columns: 12,
            mode: 'flex'
        },
        media: {
            wide: 'only screen and (min-width: 769px)',
            narrow: 'only screen and (max-width: 768px)'
        }
    }, opt);
    css.use(require('postcss-import'));
    if (opt.modules.indexOf('at-style') >= 0) { css.use(require('postcss-at-style')); }
    if (opt.modules.indexOf('remove-comments') >= 0) { css.use(require('./modules/remove-comments.js')); }
    if (opt.modules.indexOf('media') >= 0) { css.use(require('./modules/media.js')(opt.media)); }
    if (opt.modules.indexOf('color') >= 0) { css.use(require('./modules/color')); }
    if (opt.modules.indexOf('var') >= 0) { css.use(require('./modules/var.js')); }
    if (opt.modules.indexOf('google-color') >= 0) { css.use(require('postcss-google-color')); }
    if (opt.modules.indexOf('icon') >= 0) { css.use(require('./modules/icon.js')); }

    css.use(require('postcss-node-sass')(opt.sass));

    if (opt.modules.indexOf('font') >= 0) { css.use(require('postcss-font')); }
    if (opt.modules.indexOf('font-vsize') >= 0) { css.use(require('postcss-font-vsize')); }
    if (opt.modules.indexOf('sides') >= 0) { css.use(require('postcss-sides')); }
    if (opt.modules.indexOf('timing-function') >= 0) { css.use(require('postcss-timing-function')); }
    if (opt.modules.indexOf('clear-fix') >= 0) { css.use(require('postcss-clear-fix')); }
    if (opt.modules.indexOf('morphicon') >= 0) { css.use(require('postcss-morphicon')); }
    if (opt.modules.indexOf('gridder') >= 0) { css.use(require('postcss-gridder')(opt.grid)); }
    return css;
});

modules/var.js
let postcss = require('postcss'),
    util = require('postcss-plugin-utilities');

module.exports = postcss.plugin('postcss-styler-var', () => {
    let applyVars = o => {
        ['params', 'selector', 'prop', 'value'].forEach(prop => {
            if (o.hasOwnProperty(prop)) {
                o[prop] = o[prop].replace(/(^v\(|[\s\(\,\)]v\()/ig, (s) => { return s.replace('v(', 'var('); });
                o[prop] = o[prop].replace(/(^c\(|[\s\(\,\)]c\()/ig, (s) => { return s.replace('c(', 'var(color-'); });
                o[prop] = o[prop].replace(/var\(([^\,\)]+)(\,((([^\(\)]+)?\(([^\(\)]+)?\)([^\(\)]+)?)|[^\)]+))?\)/ig, (s, v, d = null) => {
                    if (d) { d = d.slice(2); }
                    if (d === 'null' || d === 'false') { d = null; }
                    if (util.sassHasVar(v, o)) { return `$${v}`; }
                    else if (d) { return d; }
                    else {
                        if (o.type !== 'atrule') {
                            removeEmpty(o);
                        }
                        return null;
                    }
                });
            }
        });
    };
    let removeEmpty = o => {
        let parent = o.parent;
        o.remove();
        if (parent && parent.nodes.lenght == 0) {
            removeEmpty(parent);
        }
    }
    return css => {
        css.walkAtRules(atrule => { applyVars(atrule); });
        css.walkRules(rule => { applyVars(rule); });
        css.walkDecls(decl => { applyVars(decl); });
    }
});

modules/color.js
let postcss = require('postcss'),
    util = require('postcss-plugin-utilities');

module.exports = postcss.plugin('postcss-styler-color', () => {
    let applyColor = o => {
        ['params', 'selector', 'prop', 'value'].forEach(prop => {
            if (o.hasOwnProperty(prop)) {
                o[prop] = o[prop].replace(/(^c\(|[\s\(\,\)]c\()/ig, (s) => { return s.replace('c(', 'color-variable('); });
                o[prop] = o[prop].replace(/color\-variable\(([^\s\)]+)(\s((([^\(\)]+)?\(([^\(\)]+)?\)([^\(\)]+)?)|[^\)]+))?\)/ig, (s, v, d = null) => {
                    let color = util.sassGetVar(`color-${v}`, o);
                    if (color === null) {
                        util.removeNode(o);
                    } else {
                        if (d !== null) {
                            d = d.trim();
                            if (color.indexOf('google-color(') >= 0) {
                                if (d === '95%') { d = '50'; }
                                if (d === '90%') { d = '100'; }
                                if (d === '80%') { d = '200'; }
                                if (d === '70%') { d = '300'; }
                                if (d === '60%') { d = '400'; }
                                if (d === '50%') { d = '500'; }
                                if (d === '40%') { d = '600'; }
                                if (d === '30%') { d = '700'; }
                                if (d === '20%') { d = '800'; }
                                if (d === '10%') { d = '900'; }
                                if (['50', '100', '200', '300', '400', '500', '600', '700', '800', '900', 'A100', 'A200', 'A400', 'A700'].indexOf(d) >= 0) {
                                    color = color.replace(/google\-color\(([a-z]+)(\s?\,\s?[a-z0-9]+)?\)/ig, `google-color($1, ${d})`);
                                }
                            } else {
                                if (d === '50') { d = '95%'; }
                                if (d === '100' || d === 'A100') { d = '90%'; }
                                if (d === '200' || d === 'A200') { d = '80%'; }
                                if (d === '300') { d = '70%'; }
                                if (d === '400' || d === 'A400') { d = '60%'; }
                                if (d === '500') { d = '50%'; }
                                if (d === '600') { d = '40%'; }
                                if (d === '700' || d === 'A700') { d = '30%'; }
                                if (d === '800') { d = '20%'; }
                                if (d === '900') { d = '10%'; }
                                if (d.indexOf('%') >= 0) {
                                    color = `change-color($color-${v}, $lightness: ${d})`;
                                }
                            }
                        } else {
                            color = `$color-${v}`;
                        }
                        return color;
                    }
                });
            }
        });
    }
    return css => {
        css.walkAtRules(atrule => { applyColor(atrule); });
        css.walkRules(rule => { applyColor(rule); });
        css.walkDecls(decl => { applyColor(decl); });
    }
});

modules/media.js
let postcss = require('postcss'),
    util = require('postcss-plugin-utilities');

module.exports = postcss.plugin('postcss-styler-media', (opt = {}) => {
    return css => {
        opt = Object.assign({
            wide: 'only screen and (min-width: 769px)',
            narrow: 'only screen and (max-width: 768px)'
        }, opt);
        css.walkAtRules('media', atrule => { 
            let params = atrule.params.split(' and '),
                newParams = [];
            params.forEach(param => {
                if (util.sassHasVar(`media-${param}`, atrule.nodes[0])) {
                    param = `#{$media-${param}}`;
                } else if (opt.hasOwnProperty(param)) { 
                    param = opt[param];
                }
                newParams.push(param);
            });
            atrule.params = newParams.join(' and ');
        });
    }
});

modules/icon.js
let postcss = require('postcss'),
    util = require('postcss-plugin-utilities');

module.exports = postcss.plugin('postcss-styler-icon', () => {
    return css => {
        css.walkDecls('icon', decl => {
            let parent = decl.parent,
                declVal = (decl.value.match(/(?!\B(\(\")[^\"\(\)]*),(?![^\"\(\)]*(\)|\")\B)/ig) === null) ? postcss.list.space(decl.value) : postcss.list.comma(decl.value),    
                settings = util.filterObject((decl.value.indexOf(',') === -1) ? postcss.list.space(decl.value) : postcss.list.comma(decl.value), {
                    'icon': [() => { return true; }],
                    'selector': ['after', 'before'],
                    'font-size': [util.isSize],
                    'font-style': ['normal', 'italic', 'oblique', 'initial', 'inherit'],
                    'font-weight': ['100', '200', '300', '400', '500', '600', '700', '800', '900', 'normal', 'bold', 'bolder', 'lighter', 'initial', 'inherit'],
                    'line-height': [util.isSize, util.isNumber],
                    'text-align': ['left', 'right', 'center', 'justify', 'initial', 'inherit'],
                    'font-variant': ['normal', 'small-caps', 'initial', 'inherit'],
                    'color': [util.isColor],
                    'text-decoration': ['none', 'underline', 'overline', 'line-through', 'initial', 'inherit'],
                    'text-transform': ['none', 'capitalize', 'uppercase', 'lowercase', 'initial', 'inherit'],
                    'vertical-align': ['baseline', 'isSize', 'isSizeNegative', 'sub', 'sup', 'super', 'top', 'text-top', 'middle', 'bottom', 'text-bottom', 'initial', 'inherit'],
                    'word-wrap': ['normal', 'break-word', 'initial', 'inherit'],
                    'letter-spacing': [util.isSize, 'normal', 'initial', 'inherit'],
                    'word-spacing': [util.isSize, 'normal', 'initial', 'inherit'],
                    'text-indent': [util.isSize, 'initial', 'inherit'],
                    'white-space': ['normal', 'nowrap', 'pre', 'pre-line', 'pre-wrap', 'initial', 'inherit'],
                    'font-stretch': ['normal', 'ultra-condensed', 'extra-condensed', 'condensed', 'semi-expanded', 'expanded', 'extra-expanded', 'ultra-expanded', '', 'unset', 'initial', 'inherit'],
                    'direction': ['ltr', 'rtl', 'initial', 'inherit'],
                    'unicode-bidi': ['normal', 'embed', 'bidi-override', 'initial', 'inherit']
                }, { selector: 'before' });
            if (settings.icon.indexOf('$') === 0) { settings.icon = util.sassGetVar(settings.icon, decl); }
            if (settings.icon && util.sassHasVar(`icon-${settings.icon}`, decl)) { 
                let fontName = 'icon';
                settings.icon.split('-').forEach(segment => { 
                    if (util.sassHasVar(`${fontName}-${segment}-font-family`, decl)) { 
                        fontName = `${fontName}-${segment}`;
                    }
                });
                fontName = (util.sassHasVar(`${fontName}-font-family`, decl)) ? `$${fontName}-font-family` : null;
                if (fontName) { 
                    let rule = postcss.rule();
                    if (parent.type === 'atrule') { 
                        rule.selector = `&:${settings.selector}`;
                    } else {
                        rule.selector = util.eachSelector(parent.selector, `&:${settings.selector}`);
                    }                    
                    rule.append({ prop: 'font-family', value: fontName });
                    rule.append({ prop: 'content', value: `'#{$icon-${settings.icon}}'` });
                    delete settings.selector; delete settings.icon;
                    for (let [prop, val] of Object.entries(settings)) { 
                        rule.append({ prop: prop, value: val });
                    }
                    parent.before(rule);
                }
            }
            decl.remove();
        });
    }
});